### ISSUE TITLE
Service Account Key Rotation

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#20165]
A user asked how to rotate service account keys.
[discussion thread](https://groups.google.com/forum/#!msg/google-containers/qSo2Q5kGI6I/use7PIYuCAAJ)

Currently, user restarts controller manager with new keypair and then restarts all pods.
User pointed out problems with that procedure:

1. you cannot restart pod instantly
2.  until a pod is restarted, it is basically broken when assuming it requires access to the API.  (KubeDNS in particular is a problem)
3. many pods are restarted which actually don't have to be restarted as they don't access the API

For problem 2, I suggested the following improved procedure:

1. Cluster is started with keypair 1
  - `kube-apiserver -service-account-key-file=pub-1` 
  - `kube-controller-mgr --service-account-private-key-file=priv-1`.
2. time passes.  service account tokens are generated with first keypair
3. it is time to rotate.  generate new key pair, `pub-2` and `priv-2`
5. run with both keypairs
  -  Accept both: `kube-apiserver -service-account-key-file=pub-1,pub-2`
  - Generate new: `kube-controller-mgr --service-account-private-key-file=priv-2`
6. restart pods at a leisurely pace.
7. stop accepting old tokens
  -  `kube-apiserver -service-account-key-file=pub-1` 
  - `kube-controller-mgr --service-account-private-key-file=priv-1`

What is missing is that ` -service-account-key-file` only accepts a single public key.  Code deeper in the token authenticator appears to handle a list of keys.  

For problem 3:

Action items:

- [ ] make the flag take a list
- [ ] close #16779 so that it can be explicit which pods do not need (and so do not have) a service account token.
- [ ] document rotation procedure, mentioning how to find and to skip pods without a token.
- [ ] e2e test where keys are rotated

**Comment by erictune** [comment]
@liggitt since it is about service accounts

**Comment by erictune** [comment]
@mml since it is about restarting all the pods.

**Comment by liggitt** [comment]
yeah, the token generation/validation tolerates rotation... plumbing a list of keys (newest to oldest) should be sufficient from the authenticator's perspective.

**Comment by erictune** [comment]
Edited procedure above to add missing step: need to delete old secrets and add new secrets to each service account prior to restarting pods.

**Comment by redbaron** [comment]
it is all nice in theory, but knowing when doing step `7` is safe is not a trivial task, or at least I couldn't figure it out.

If instead k8s client libraries (which I believe is now called `client-go`) were able to reload service account token when it gets updates on a filesystem, then all the dance with pod restarts will be not necessary.

The procedure will be greatly simplified:
1. generate new keys
2. restart apiserver and controller manager, pass both old and new keys to apiserver
3. `kubectl patch` all service account secrets to delete `data.token`, then because of https://github.com/kubernetes/kubernetes/blob/6079053407705212c8cc3f039e8b767c6bccd0a0/pkg/controller/serviceaccount/tokens_controller.go#L464 all keys will be regenerated *under the same secret name*, which would trigger update on a filesystem in every pod using them
4. [missing part] k8s library detects new token file in "/var/run/secrets/kubernetes.io/serviceaccount/token" which was helpfully updated by kubelet and starts using for new requests

**Comment by liggitt** [comment]
> kubectl patch all service account secrets to delete data.token, then because of https://github.com/kubernetes/kubernetes/blob/6079053407705212c8cc3f039e8b767c6bccd0a0/pkg/controller/serviceaccount/tokens_controller.go#L464 all keys will be regenerated under the same secret name, which would trigger update on a filesystem in every pod using them

that part is the disruptive part... when rotating, you want to be able to have a period where both the old and new tokens are valid, but the new token is preferred/used by new pods. patching out the old token revokes it.

**Comment by redbaron** [comment]
why? you still run  apiserver with both old and new keys, so that even if there is a time period where k8s client libraries continue to use old key it still works. net effect is that you don't  identify  pods which need a rolling restart and neither you need to perform that restart at all.

**Comment by liggitt** [comment]
Servers that allow service account token revocation run with `--service-account-lookup`, which verifies the token in etcd matches

**Comment by redbaron** [comment]
Ah, didn't think about that.So whats the solution? Rolling all the pods isn't something which one to do lightly

**Comment by ericchiang** [comment]
> where all apiservers and controllermanagers need to have the same serviceaccount keypair.

@johanot they don't. API servers can already be given multiple public keys to validate service account JWTs (https://github.com/kubernetes/kubernetes/pull/34029). This lets you introduce a new key pair over time for validation before changing it in the controller manager as the active signing key.

We have longer term plans for automatically rotating service accounts credentials https://github.com/kubernetes/community/pull/1460

Until then, this issue is just about document the flags you need to update when changing the key.

**Comment by kfox1111** [comment]
+1 for sticking it in etcd. I think if someone is able to compromise etcd, they can stick whatever they want into the cluster already. Far worse then getting the key. So keeping the service account keys there shouldn't really carry any additional risk, but has a lot of benefits to simplifying rotation.

**Comment by johanot** [comment]
@kfox1111 I'd argue to only share _public_ keys in etcd, for the apiservers to agree on which signers to trust. Private keys can be generated locally at each master node and kept there, in tmpfs even, because if they get deleted, a new keypair can just be generated. 

Thus; Neither having read access to etcd nor access to data at rest on disk gives you any power. Only by gaining _write_ access to etcd, you will be put in control of altering the list of what to trust.

If you really want missile-key kind of security, you could combine this pattern with regular PKI chain-of-trust and provide the apiserver with a ca-certificate on command line, by which each service account cert must be signed. Then you'd need access to both the CA-key and etcd, or access to change apiservers at cmdline to trust another CA. However, kind of the same security can be gained by just securing etcd with rbac and client cert auth.

**Comment by mikedanese** [comment]
Ref https://github.com/kubernetes/enhancements/pull/704

**Comment by 2rs2ts** [comment]
What about rotating the CA for the signing keys? Has that been thought through?

**Comment by liggitt** [comment]
the service account keys are simple JWTs, verified by a set of public keys; there's no CA involved

**Comment by kfox1111** [comment]
> the service account tokens are simple JWTs, and are verified by a set of public keys, so the signing keys are just keys; there's no CA involved

True, but still could be useful to allow updating the signing keys?

**Comment by jkevlin** [comment]
What is the status of this?  Is there a documented procedure for rotating service account credentials?

**Comment by ritazh** [comment]
/assign @micahhausler 
to provide link to the KEP and add docs for this feature

**Comment by mattmoyer** [comment]
Prior to step 7 (stop accepting old tokens), it would also be nice to observe a kube-apiserver metric that shows how many active requests are being authenticated with each of the public keys.

You could use this to confirm that no stale tokens are still being validated under an older key before you remove it. Public keys would need to be identified somehow -- perhaps by fingerprint?

**Comment by fiksn** [comment]
It is true that you can't easily rotate key, but as far as I have checked you can still rotate service account secrets. Trick is that JWT contains something like:

```"kubernetes.io/serviceaccount/secret.name": "default-token-g8rln",```

so while the JWT has no expiry and will remain valid forever, API server checks for the existence of `default-token-g8rln`. As soon as I've deleted this secret (which caused the creation of a new one for default service account) I wasn't able to access the API with the old JWT anymore. Of course if private key gets compromised this doesn't help you, but if just the k8s secrets were, you are good.

**Comment by sftim** [comment]
Is there a linked issue against k/website to document the rotation process as a task? (if not, I recommend opening one)

**Comment by andreyklumchyk** [comment]
Here is my way to k8s certs rotation without downtime.

First of all, I do not change CA certificate because it has a much longer lifetime
also, there will need a little bit of complex instruction and I am just working on it.

P.S. before making any movement try on some sandbox.

1. Preparation.

As I told you before do not change CA key and cert. Also You will be needed a copy of ServiceAccount cert on all master nodes.
Next, we need to deploy some temporal changes to the `kube-apiserver` service.
```
kube-apiserver
  ...
  --service-account-key-file={{ k8s_tls_dir }}/service-account-old.crt
  ...
  --anonymous-auth=true
```

2. Renew all other keys and certs (certs must be signed with old CA)
```
openssl genrsa -out admin.key 2048
openssl req -new -key admin.key -out admin.csr \
        -subj "/CN=admin/O=system:masters"
openssl x509 -req -days 365 \
        -CA ca.crt -CAkey ca.key -CAcreateserial \
        -in admin.csr -out admin.crt
```

3. Copy all keys and certs to the hosts

4. Reload `kubelet` service on all hosts

5. Delete all `kube-apiserver-*` and `kube-controller-manager-*` pods on master nodes

6. Delete `default` ServiceAccount on all namespaces
```
kubectl delete sa default
```

7. Delete all Secrets of ServiceAccounts

```
kubectl get secrets \
  --field-selector type=kubernetes.io/service-account-token \
  -o=jsonpath='{.items[*].metadata.name}'

kubectl delete secret <names from pref step> --ignore-not-found=true
```

8. Done. Now You could roll back changes in the `kube-apiserver` service.
